Numpy reader C++ changes

SundarRajan28 opened the pull request.

Commit:
Changes in file_source_reader - to minimize the I/O operations

Commit:
Changes in the variable name

Commit:
Changes in the variable names of the audio source evalution

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

swetha097 committed:
Use set instead of vector

swetha097 committed:
Minor bug fixes

fiona-gladwin committed:
Minor fixes

Remove changes to update the
filenames vector incase of padding
Fix the pipeline when pad is off and use idx in a continuous manner

fiona-gladwin committed:
Fix drop policy without padding

To skip and batch and start with the previously padded idx in last batch

fiona-gladwin committed:
Fix pytorch iterator - PARTIAL policy

fiona-gladwin committed:
Merge branch 'swbs/audio/pr10' of https://github.com/swetha097/rocAL into lbp_fix_pr10

SundarRajan28 committed:
Merge branch 'audio_pr4' into swbs/audio/pr5

SundarRajan28 committed:
Merge branch 'swbs/audio/pr5' into swbs/audio/pr6

SundarRajan28 committed:
Merge branch 'swbs/audio/pr6' into swbs/audio/pr8

SundarRajan28 committed:
Merge branch 'swbs/audio/pr8' into swbs/audio/pr9

SundarRajan28 committed:
Merge branch 'swbs/audio/pr9' into swbs/audio/pr7

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr2

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into swbs/audio/pr3

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'audio_pr4' into swbs/audio/pr5

SundarRajan28 committed:
Merge branch 'swbs/audio/pr5' into swbs/audio/pr6

SundarRajan28 committed:
Merge branch 'swbs/audio/pr6' into swbs/audio/pr8

SundarRajan28 committed:
Merge branch 'swbs/audio/pr8' into swbs/audio/pr9

SundarRajan28 committed:
Merge branch 'swbs/audio/pr9' into swbs/audio/pr7

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into swbs/audio/pr1

fiona-gladwin committed:
Merge branch 'swbs/audio/pr1' into swbs/audio/pr2

fiona-gladwin committed:
Revert empty line removed in CMakeLists.txt

fiona-gladwin committed:
Removed prefix original for audio vectors

fiona-gladwin committed:
Fix PARTIAL

Fix issue with batch size greater than dataset size

swetha097 committed:
Reduce overall time for audio source evalution

swetha097 committed:
Fix shard_size and stick to shard issue seen with convergence

swetha097 committed:
Resolve PR comments

swetha097 committed:
Add @params to all args in pytorch.py

swetha097 committed:
Fix build issue

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into swbs/audio/pr2

fiona-gladwin committed:
Merge branch 'swbs/audio/pr7' of https://github.com/swetha097/rocAL into swbs/audio/pr10

swetha097 committed:
Minor changes in unit test

swetha097 committed:
Merge branch 'swbs/audio/pr2' of https://github.com/swetha097/rocAL into swbs/audio/pr2

swetha097 committed:
Minor changes

swetha097 committed:
Change ROCAL instaces to rocAL in pytorch.py

fiona-gladwin committed:
Merge branch 'swbs/audio/pr2' into swbs/audio/pr3

swetha097 committed:
Resolve the PR comments

swetha097 committed:
Merge branch 'swbs/audio/pr3' of https://github.com/swetha097/rocAL into swbs/audio/pr3

swetha097 committed:
Minor changes in decoders.py - Modify the comment for shard_size

fiona-gladwin committed:
Merge branch 'swbs/audio/pr2' of https://github.com/swetha097/rocAL into swbs/audio/pr3

fiona-gladwin committed:
Merge branch 'swbs/audio/pr3' of https://github.com/swetha097/rocAL into swbs/audio/pr3

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into swbs/audio/pr3

swetha097 committed:
Merge branch 'swbs/audio/pr10_training' into swbs/audio/pr10

swetha097 committed:
Fix shard_size

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

swetha097 committed:
Minor changes

swetha097 committed:
Changes in pipeline.py and decoders.py

swetha097 committed:
Merge remote-tracking branch 'origin/swbs/audio/pr10_training' into swbs/audio/pr10

swetha097 committed:
Address the PR comments

swetha097 committed:
Address Review comments

swetha097 committed:
Remove print statement

swetha097 committed:
Merge branch 'swbs/audio/pr10' into lbp_fix_pr10

swetha097 committed:
Fix the count_items

swetha097 committed:
Make Sharding similar to DALI

swetha097 committed:
Fix issues with DROP policy by introducing a new vector for padding

swetha097 committed:
Minor fixes

swetha097 committed:
Comment out print statements

swetha097 committed:
Add changes for shard_size LBP testing

swetha097 committed:
Fix DROP Policy with shard_size > 0

swetha097 committed:
Fix Stick_to_Shard=False

swetha097 committed:
Fix PARTIAL policy and code clean up

swetha097 committed:
fix last_batch_padded size when shard_size > 0

swetha097 committed:
Fix Drop policy - we skip the dropped batch in the next epoch

swetha097 committed:
Fix single shard outputs

swetha097 committed:
Remove the commented code and fix the padding code in open()

swetha097 committed:
Remove div by num_shards in decoders.py

fiona-gladwin committed:
Introduce Audio layouts

fiona-gladwin committed:
Add layout changes for spectrogram

swetha097 committed:
Fix the unit tests - c++ & python

fiona-gladwin committed:
Merge branch 'swbs/audio/pr5' of https://github.com/swetha097/rocAL into swbs/audio/pr5_layout

swetha097 committed:
Code clean up and formatting

swetha097 committed:
Minor code clean up

swetha097 committed:
code clean up in pytorch.py

fiona-gladwin committed:
Add layout changes for spectrogram

fiona-gladwin committed:
Pass layouts for MelFilterBank

fiona-gladwin committed:
Fix ToDecibels

Pass layouts for ToDecibels

fiona-gladwin committed:
Fix Normalize

fiona-gladwin committed:
Fix build issue

fiona-gladwin committed:
Merge branch 'swbs/audio/pr5_layout' of https://github.com/swetha097/rocAL into swbs/audio/pr7_layout

fiona-gladwin committed:
Fix python unit test

swetha097 committed:
Merge remote-tracking branch 'swe_fork/swbs/audio/pr7_layout' into swbs/lbp_fixes_pr10

fiona-gladwin committed:
Minor fix

swetha097 committed:
Pass LBP to decoders instead of the Pipeline creation

Commit:
Merge branch 'swbs/lbp_fixes_pr10_pass_readers' into swbs/lbp_fixes_pr10

swetha097 committed:
Update pipeline.py - Remove commented code

swetha097 committed:
Update pipeline.py - Remove commented out code

SundarRajan28 committed:
Adding changes for spec layout changes

SundarRajan28 committed:
Merge branch 'swbs/audio/pr6' into swbs/audio/pr8

SundarRajan28 committed:
Merge branch 'swbs/audio/pr8' into swbs/audio/pr7

SundarRajan28 committed:
Adding changes to MFB and normalize nodes

swetha097 committed:
Update node_slice.cpp

swetha097 committed:
Update node_slice.h

swetha097 committed:
Resolve PR comments

fiona-gladwin committed:
Merge branch 'swbs/audio/pr5_layout' into swbs/audio/pr5

fiona-gladwin committed:
Merge branch 'swbs/audio/pr5' into swbs/audio/pr6

fiona-gladwin committed:
Merge branch 'swbs/audio/pr6' into swbs/audio/pr8

fiona-gladwin committed:
Merge branch 'swbs/audio/pr8' into swbs/audio/pr9

fiona-gladwin committed:
Merge branch 'swbs/audio/pr9' into swbs/audio/pr7

swetha097 committed:
Merge remote-tracking branch 'open_source/develop' into swbs/audio/pr3

swetha097 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

swetha097 committed:
Merge branch 'audio_pr4' into swbs/audio/pr5

swetha097 committed:
Merge remote-tracking branch 'origin/swbs/audio/pr5' into HEAD

swetha097 committed:
Merge branch 'temp_swbs/audio/pr6' into swbs/audio/pr6

swetha097 committed:
Merge branch 'swbs/audio/pr6' into swbs/audio/pr8

swetha097 committed:
Merge branch 'swbs/audio/pr8' into swbs/audio/pr9

swetha097 committed:
Merge branch 'swbs/audio/pr9' into swbs/audio/pr7

swetha097 committed:
Merge branch 'swbs/audio/pr7' into swbs/lbp_fixes_pr10

swetha097 committed:
Fix downmix failing case and resolve the issue with merge

swetha097 committed:
Fix issue with file_source_reader.cpp when file_list is not used

swetha097 committed:
Add the changes for LBP in coco-file-source-reader.

swetha097 committed:
Resolve PR comments - Sundar

swetha097 committed:
Fix undefined reference

swetha097 committed:
Fix file_source_reader.cpp

SundarRajan98 committed:
Fix shuffle issues

swetha097 committed:
Merge remote-tracking branch 'swe_fork/swbs/lbp_fixes_pr10' into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Fix coco_file_source

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
make stick_to_shard as true by default

swetha097 committed:
Fix minor bug

swetha097 committed:
Change last_patch_padded to pad_last_batch_repeated

swetha097 committed:
Fix the shuffle with aspect ratios

swetha097 committed:
Make a minor change in the coco-file-source - sorting is not required

swetha097 committed:
Another fix for shuffle with aspect_ratios

swetha097 committed:
Another fix for shuffle with aspect_ratios

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Modify the split_count

swetha097 committed:
Fix name of shard_count

swetha097 committed:
Adding print statements

swetha097 committed:
Add the sorting back

swetha097 committed:
Fix the indexs calc

swetha097 committed:
Fix the indexs calc

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Add some debug statements

swetha097 committed:
Fix the ARG

swetha097 committed:
comment out part of ASR

swetha097 committed:
Comment out shuffle_with_asr

swetha097 committed:
Uncomment Line 234 & 236

swetha097 committed:
Uncomment Line 234 & 236

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10_image_readers' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10_image_readers

swetha097 committed:
Uncomment the ARG

swetha097 committed:
Add Cmake changes for HIP backend

swetha097 committed:
Add support for TF record reader

SundarRajan28 committed:
Merge branch 'develop' into numpy_reader

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr7

swetha097 committed:
Adding changes for lbp & sharding in caffe reader

swetha097 committed:
Adding changes for lbp & sharding in caffe2 lmdb reader

swetha097 committed:
Adding changes for lbp & sharding in cifar10 reader

swetha097 committed:
Adding changes for lbp & sharding in mxnetIO reader

swetha097 committed:
Minor change in tf record reader

SundarRajan28 committed:
Adding partial LBP changes to np reader

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into numpy_reader

SundarRajan28 committed:
Merge branch 'swbs/audio/pr7' into numpy_reader

sbavasab committed:
Adding comments to all if conditions

swetha097 committed:
Adding the lbps function implementation

swetha097 committed:
Minor change

swetha097 committed:
Fix a bug in the caffe2 reader

SundarRajan28 committed:
Merge remote-tracking branch 'sw_dev/swbs/lbp_fixes_pr10_image_readers' into numpy_reader

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/lbp_fixes_pr10_image_readers

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into numpy_reader

SundarRajan28 committed:
Minor changes

SundarRajan28 committed:
Merge remote-tracking branch 'sw_dev/swbs/lbp_fixes_pr10_image_readers' into numpy_reader

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10

SundarRajan28 committed:
Fix count_items

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr7

swetha097 committed:
Merge remote-tracking branch 'open_source/develop' into swbs/audio/pr9

SundarRajan28 committed:
Fix merge conflicts

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr7

SundarRajan28 committed:
Resolving review comments

swetha097 committed:
Merge remote-tracking branch 'open_source/develop' into develop

swetha097 committed:
Merge remote-tracking branch 'open_source/develop' into swbs/audio/pr9

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr7

SundarRajan28 committed:
Resolving review comments

swetha097 committed:
Merge branch 'develop' into swbs/audio/pr9

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr9

SundarRajan28 committed:
Merge branch 'swbs/audio/pr9' into swbs/audio/pr7

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/audio/pr7

swetha097 committed:
Merge remote-tracking branch 'swe_fork/swbs/audio/pr7' into swbs/lbp_fixes_pr10

swetha097 committed:
Fix a minor warning in file source reader

LakshmiKumar23 committed:
Merge branch 'develop' into swbs/audio/pr7

SundarRajan28 committed:
Resolving review comments

swetha097 committed:
LBP comments resolution

LakshmiKumar23 committed:
Merge branch 'develop' into swbs/audio/pr7

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into swbs/lbp_fixes_pr10

SundarRajan28 committed:
Merge branch 'swbs/audio/pr7' into swbs/lbp_fixes_pr10

SundarRajan28 committed:
Formatting changes

swetha097 committed:
Resolving Final Set of PR comments

swetha097 committed:
Combine with OR condition

swetha097 committed:
Remove the pad_last_batch_repeated print statement from decoders.py

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into swbs/lbp_fixes_pr10

swetha097 committed:
Add shard_size and stick_to_shard variables in args

swetha097 committed:
Merge branch 'develop' into swbs/lbp_fixes_pr10

swetha097 committed:
Minor spelling fix

kiritigowda committed:
Merge branch 'develop' into swbs/lbp_fixes_pr10

swetha097 committed:
Make changes to insert the padded data in the file_names vector

swetha097 committed:
Merge branch 'swbs/lbp_fixes_pr10' of https://github.com/swetha097/rocAL into swbs/lbp_fixes_pr10

swetha097 committed:
Support to pass the variables fo lbp as struct

swetha097 committed:
Fix segmentation fault

SundarRajan28 committed:
Merge branch 'ROCm:develop' into swbs/lbp_fixes_pr10_pr_comments

SundarRajan28 committed:
Merge branch 'numpy_reader' into unet_integration

SundarRajan28 committed:
Removing PR11 changes

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into numpy_reader

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into numpy_reader

SundarRajan28 committed:
Adding LBP changes to numpy reader

SundarRajan28 committed:
Adding i16 support to rocAL and np reader

SundarRajan28 committed:
Add npy version check

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into numpy_reader

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Adding numpy loader changes

SundarRajan28 committed:
Change copyright year

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into numpy_loader_changes

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Resolve review comments

SundarRajan28 committed:
Merge branch 'develop' into numpy_loader_changes

kiritigowda committed:
Merge branch 'develop' into numpy_loader_changes

kiritigowda committed:
Merge branch 'develop' into numpy_loader_changes

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into numpy_reader

SundarRajan28 committed:
Merge branch 'numpy_loader_changes' into numpy_reader

SundarRajan28 committed:
Fixing issue with setting dtype

SundarRajan28 committed:
Minor changes

SundarRajan28 committed:
Merge branch 'numpy_reader' into numpy_reader_cpp_changes

SundarRajan28 committed:
Adding ROI changes for numpy reader

SundarRajan28 committed:
Updating unit test for numpy reader

shobana-mcw commented on rocAL/include/api/rocal_api_data_loaders.h:
```diff
@@ -617,6 +617,58 @@ extern "C" RocalTensor ROCAL_API_CALL rocalRawTFRecordSourceSingleShard(RocalCon
+ * \param [in] rocal_sharding_info The members of RocalShardingInfo determines how the data is distributed among the shards and how the last batch is processed by the pipeline.
+ * \return Reference to the output tensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalNumpyFileSource(RocalContext p_context,
```
Rename this to context.

shobana-mcw commented on rocAL/include/api/rocal_api_data_loaders.h:
```diff
@@ -617,6 +617,58 @@ extern "C" RocalTensor ROCAL_API_CALL rocalRawTFRecordSourceSingleShard(RocalCon
+ * \param [in] rocal_sharding_info The members of RocalShardingInfo determines how the data is distributed among the shards and how the last batch is processed by the pipeline.
+ * \return Reference to the output tensor
+ */
+extern "C" RocalTensor rocalNumpyFileSourceSingleShard(RocalContext p_context,
```
Rename to context.

shobana-mcw commented on rocAL/include/pipeline/tensor.h:
```diff
@@ -137,8 +137,9 @@ class TensorInfo {
         if (_data_type == data_type)
```
Why are we checking this? 
Please check if modify strides is not needed before this return statement.

SundarRajan28 replied:
If the data_type is same as existing dtype, it just returns without modifying the strides

shobana-mcw commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    /*!
+     \param buf User's provided buffer to receive the loaded numpy data
+     \param raad_size Requested read size by the user
+     \param buf Max dimension strides of the output tensor used for padding along the dimensions
```
Rename to max_shape.

shobana-mcw commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    std::string get_root_folder_path() override;  // Returns the root folder path
+
+   private:
+    //! opens the folder containing the numpy arrays
```
Please check if all these are used in numpy data reader cpp file.

shobana-mcw commented on rocAL/source/loaders/image/numpy_loader.cpp:
```diff
@@ -154,7 +154,9 @@ NumpyLoader::load_routine() {
-    auto max_shape = _output_tensor->info().max_shape();
+    auto max_shape = _output_tensor->info().dims();
+    auto data_layout = _output_tensor->info().layout();
+    max_shape.erase(max_shape.begin());
```
Why are we erasing the shape? Please add a comment.

SundarRajan28 replied:
_output_tensor->info().dims() contains the batch dimension so removing it to get the max shape of the tensor. I can't use max_shape as it will only give (maxW, maxH) for image layouts. I have added a comment for the same

shobana-mcw commented on rocAL/source/pipeline/tensor.cpp:
```diff
@@ -259,8 +265,8 @@ void Tensor::update_tensor_roi(const std::vector<std::vector<uint32_t>> &shape)
     for (unsigned i = 0; i < info().batch_size(); i++) {
-        if (shape[i].size() != (info().num_of_dims() - 1))
-            THROW("The number of dims to be updated and the num of dims of tensor info does not match")
+        if (shape[i].size() != max_shape.size())
```
Why this has been changed?

SundarRajan28 replied:
This function was added as part of audio support. For audio, the tensor shape is NHW/NFT/NTF and we update the ROI for all dims. This condition checks whether the number of dims and the ROI shape matches. But for image layouts (NHWC/NCHW), the num_dims and the ROI shape won't match as we only store ROI for WH values.
Hence I modified the condition to check whether the ROI shape and the max_shape length is the same which holds true for both audio and image layouts

shobana-mcw reviewed (commented):
Resolve the review comments. I am yet to review numpy_reader.cpp file.

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
Change to 2025

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+*/
+
+#include <math.h>
+
```
Remove this line, ensure all imports are required here

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    std::vector<std::string> _file_names;
+    std::vector<std::string> _files;
+    std::vector<NumpyHeaderData> _file_headers;
+    FILE *_current_fPtr;
```
Please change all to camel case

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+    _file_count_all_shards = 0;
+}
+
+unsigned NumpyDataReader::count_items() {
+    int size = get_max_size_of_shard(_batch_size, _loop);
+    int ret = (size - _read_counter);
+    if (_sharding_info.last_batch_policy == RocalBatchPolicy::DROP && _last_batch_padded_size != 0)
+        ret -= _batch_size;
+    return ((ret < 0) ? 0 : ret);
+}
```
Check the count items in imag reader.cpp in other files we use the count items present in that

SundarRajan28 replied:
This change is part of C++ test suite fix PR. I will merge the changes once its pulled into ToT

fiona-gladwin replied:
@SundarRajan28 Is this done?

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    unsigned _seed = 0;
```
loop, shuffle, seed, 
All are present as part of the Reader class
These changes are part of the C++ test suite fix PR
We might need to merge that before this PR, so we can have the Reader changes from that PR

SundarRajan28 replied:
I will merge the changes once its pulled into ToT

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    std::map<std::string, NumpyHeaderData> _header_cache;
+    const RocalTensorDataType get_dtype(const std::string& format);
+    inline void skip_spaces(const char*& ptr);
+    void parse_header_data(NumpyHeaderData& target, const std::string& header);
```
Can we have a comment for these functions to elaborate it's usage

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    void update_header_cache(const std::string& file_name, const NumpyHeaderData& value);
+    void incremenet_read_ptr();
+    int release();
+    std::shared_ptr<MetaDataReader> _meta_data_reader = nullptr;
```
Move this to the beginning where we have declared all other variables

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+
+class NumpyDataReader : public Reader {
+   public:
+    //! Looks up the folder which contains the files, amd loads the numpy files
```
It is and here not amd

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+    Reader::Status initialize(ReaderConfig desc) override;
+    //! Reads the next resource item
+    /*!
+     \param buf User's provided buffer to receive the loaded numpy data
```
define max_size param

SundarRajan28 replied:
Not used in numpy reader so ignoring param

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+
+    NumpyDataReader();
+
+    std::string get_root_folder_path() override;  // Returns the root folder path
```
Add this comment in prev line similar to other files

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+            }
+            _last_file_name = _absolute_file_path;
+        } else {
+            std::ifstream fp(_file_list_path);
```
Will we have a list file passed to numpy reader?
Do we have such use case?

SundarRajan28 replied:
For trainings, we only pass a list of file paths to numpy reader for each shard. But DALI has the option to pass file lists and since we already support it in file_source_reader, I have added it for numpy reader as well

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+                }
+            }
+        }
+    } else if (!_files.empty()) {
```
Add a comment why this is added

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+        return std::fread((unsigned char*)buf, sizeof(unsigned char), _curr_file_header.nbytes(), _current_fPtr);
+
+    if (_curr_file_header.type() == RocalTensorDataType::UINT8)
+        actual_read_size = parse_numpy_data<u_int8_t>((u_int8_t*)buf, strides_in_dims, shape);
```
What is u_int8 / u_int32

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+        THROW("Error extracting header length.");
+
+    // read header: the offset is a magic number
+    int64_t offset = 10;
```
Why is this offset used?

shobana-mcw replied:
Please add comment explaining offset use. Also what is meant by the comment magic number?

SundarRajan28 replied:
The first 6 bytes are a magic string: exactly \x93NUMPY.
The next 1 byte is an unsigned byte: the major version number of the file format, e.g. \x01.
The next 1 byte is an unsigned byte: the minor version number of the file format, e.g. \x00. Note: the version of the file format is not tied to the version of the numpy package.
The next 2 bytes form a little-endian unsigned short int: the length of the header data HEADER_LEN.
We read from that offset to get the numpy header and then decode it.

fiona-gladwin commented on rocAL/include/readers/image/numpy_data_reader.h:
```diff
@@ -0,0 +1,134 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
Change copyright

fiona-gladwin commented on rocAL/include/api/rocal_api_data_loaders.h:
```diff
@@ -617,6 +617,58 @@ extern "C" RocalTensor ROCAL_API_CALL rocalRawTFRecordSourceSingleShard(RocalCon
+ * \param [in] seed Determines the seed used by RNG for shuffling data between shards.
+ * \param [in] rocal_sharding_info The members of RocalShardingInfo determines how the data is distributed among the shards and how the last batch is processed by the pipeline.
+ * \return Reference to the output tensor
+ */
```
Rajy had suggested this earlier whichever API we are using we can retain that alone

SundarRajan28 replied:
We use this API in the C++ unit tests for numpy reader test case

fiona-gladwin replied:
But can we not use the other API in C++?
Do we need to have both please check once with Shobana if needed

fiona-gladwin commented on rocAL/include/api/rocal_api_data_loaders.h:
```diff
@@ -617,6 +617,58 @@ extern "C" RocalTensor ROCAL_API_CALL rocalRawTFRecordSourceSingleShard(RocalCon
+ * \param [in] files Contains a list of file paths to read the data from.
+ * \param [in] is_output Determines if the user wants the loaded images to be part of the output or not.
+ * \param [in] shuffle Determines if the user wants to shuffle the dataset or not.
+ * \param [in] loop Determines if the user wants to indefinitely loops through images or not.
```
Change from loops to loop

fiona-gladwin commented on rocAL/source/api/rocal_api_data_loaders.cpp:
```diff
@@ -1686,6 +1703,105 @@ rocalVideoFileSource(
+        RocalTensorlayout op_tensor_layout = static_cast<RocalTensorlayout>(output_layout);
+        std::vector<size_t> dims(max_dimensions.size() + 1);
+        dims[0] = context->user_batch_size();
+        for (uint i = 0; i < max_dimensions.size(); i++)
```
Can we not use assign or insert to insert a set of values instead of for loop?

fiona-gladwin commented on rocAL/source/api/rocal_api_data_loaders.cpp:
```diff
@@ -1686,6 +1703,105 @@ rocalVideoFileSource(
+        auto info = TensorInfo(std::vector<size_t>(std::move(dims)),
+                               context->master_graph->mem_type(),
+                               tensor_data_type);
+        info.set_tensor_layout(op_tensor_layout);
```
Can we use the other TensorInfo where we pass the layout as well?

fiona-gladwin commented on rocAL/include/pipeline/tensor.h:
```diff
@@ -349,6 +350,7 @@ class Tensor : public rocalTensor {
     int create_virtual(vx_context context, vx_graph graph);
     bool is_handle_set() { return (_vx_handle != 0); }
     void set_dims(std::vector<size_t> dims) override { _info.set_dims(dims); }
+    void set_layout(RocalTensorlayout layout) { _info.set_tensor_layout(layout); }
```
Is this used anywhere?

fiona-gladwin commented on rocAL/include/pipeline/tensor.h:
```diff
@@ -137,8 +137,9 @@ class TensorInfo {
         _data_type = data_type;
-        _data_size = (_data_size / _data_type_size);
-        _data_size *= data_type_size();
+        _data_type_size = tensor_data_size(_data_type);
```
Is this needed? Are we modifying data type at a later point
The TensorInfo itself will update the stride right?

SundarRajan28 replied:
We found this bug during the cosmoflow integration. Log1p kernel accepts int16 inputs but returns f32 outputs. We use the input tensors info but modify the data_type using this function. Without calling modify_strides, the strides were wrong

fiona-gladwin commented on rocAL/source/loaders/image/numpy_loader.cpp:
```diff
@@ -183,7 +185,14 @@ NumpyLoader::load_routine() {
+                auto original_roi = _reader->get_numpy_header_data().shape();
+                if (data_layout == RocalTensorlayout::NHWC) {
+                    _tensor_roi[file_counter] = {original_roi[1], original_roi[0]};
+                } else if (data_layout == RocalTensorlayout::NCHW) {
```
Better to add comment to explain why this is done

shobana-mcw commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1))
+        THROW("Found wrong symbol during parsing");
```
Should we terminate execution or should it continue with just throwing the error?

SundarRajan28 replied:
In image loader, if the file headers are not readable, we return 0 as the file_size in open(). During decoding, we get another file from the same batch and substitute it for the failed file. Here, there is no decoding so I'm throwing errors even if one of the npy files failed header parsing

SundarRajan28 replied:
I will check if we can instead print errors and continue execution

shobana-mcw commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+}
+
+template <typename T>
+T NumpyDataReader::parse_int(const char*& ptr) {
```
Do we have only one function call for this? If so can we add as inline or add it in the respective function itself?

SundarRajan28 replied:
I checked online and found that templates are implicitly inlined

shobana-mcw commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,587 @@
+    return value;
+}
+
+std::string NumpyDataReader::parse_string(const char*& input, char delim_start, char delim_end) {
```
Same as previous comment.

SundarRajan28 replied:
I checked online and found that templates are implicitly inlined

shobana-mcw reviewed (changes requested):
Be on the lookout for following 
1. There are many parse functions which are called once. Check if they can be inlined or can we add it in the function itself. 
2. There are many THROW statements. 
        a. Check where its handled
        b. Add more details wrt file name(numpy reader) and function name for tracing back. 
        c. rocAL THROW will not exit the execution. Check if any THROW needs to exit and handle  it correctly. 
3. Check if the generic file reading / file parsing is required for numpy reader.

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Adding function comments

SundarRajan28 committed:
Adding checks for numpy header parsing

SundarRajan28 committed:
Adding fixes for checks

SundarRajan28 committed:
Merge branch 'develop' into numpy_reader_cpp_changes

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,672 @@
+    skip_spaces(hdr);
+    CALL_AND_CHECK_FLAG(skip_char(hdr, "{"));
+    CALL_AND_CHECK_FLAG(skip_field(hdr, "descr"));
+    typestr = parse_string(hdr);
```
You can declare here itself right, why are we declaring in line 258, same for variable little endian

SundarRajan28 replied:
That is becoz of the usage of the macro. In case it exits, the compiler was throwing errors that both typestr and little endian is undeclared. That's why I declared it and then set it only when we get to that point without any error
https://stackoverflow.com/questions/14274225/statement-goto-can-not-cross-variable-definition

fiona-gladwin commented on rocAL/source/readers/image/numpy_data_reader.cpp:
```diff
@@ -0,0 +1,672 @@
+#define CALL_AND_CHECK_FLAG(func) \
+    do { \
+        func; \
+        if (_header_parsing_failed) { \
```
Can a macro access the private variable of a class?
Have we checked this?

SundarRajan28 replied:
It is correctly setting the flag due to which it exits the open() function when I tested

SundarRajan28 committed:
Adding changes from PR267

kiritigowda committed:
Merge branch 'develop' into numpy_reader_cpp_changes

kiritigowda committed:
Merge branch 'develop' into numpy_reader_cpp_changes

SundarRajan28 committed:
Resolving review comments

LakshmiKumar23 committed:
Test - clean data paths (#298)

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Macro changes

lucbruni-amd committed:
Test - Fix README inconsistencies (#299)

* Test - Fix README inconsistencies

* Test - Fix typo

* Update tests/cpp_api/unit_tests/README.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

---------

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

SundarRajan28 committed:
Removing templated function for reading numpy data

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.17.1 to 1.18.1 in /docs/sphinx (#300)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.17.1 to 1.18.1.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.17.1...v1.18.1)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

SundarRajan28 committed:
Deps update: Support turbojpeg package (#296)

* Removing tj3 API usage

* Adding turbojpeg package in rocAL setup script

* Update changelog

* Add turbojpeg rpm pkg

---------

Co-authored-by: SundarRajan98 <svaithiy@amd.com>
Co-authored-by: Lakshmi Kumar <lakshmi.kumar@amd.com>

kiritigowda committed:
Merge branch 'develop' into numpy_reader_cpp_changes

SundarRajan28 committed:
Resolving review comments
